src/main/resources/io/viash/platforms/nextflow/workflowFactory/workflowFactory.nf

```diff
@@ -0,0 +1,206 @@
+              "Error renaming data keys in module '${key}' id '${tuple[0]}'.\n" +
+              "  Key '$oldKey' is missing in the data map. tuple[1].keySet() is '${tuple[1].keySet()}'"
+            tuple[1].put(newKey, tuple[1][oldKey])
+          }
+          tuple[1].keySet().removeAll(processArgs.renameKeys.collect{ newKey, oldKey -> oldKey })
+        }
+        tuple
+      }
+
+    if (processArgs.filter) {
+      mid2_ = mid1_
+        | filter{processArgs.filter(it)}
+    } else {
+      mid2_ = mid1_
+    }
+
+    if (processArgs.fromState) {
+      mid3_ = mid2_
+        | map{
+          def new_data = processArgs.fromState(it.take(2))
+          [it[0], new_data]
+        }
+    } else {
+      mid3_ = mid2_
+    }
+
+    // fill in defaults
+    mid4_ = mid3_
+      | map { tuple ->
+        def id_ = tuple[0]
+        def data_ = tuple[1]
+
+        // TODO: could move fromState to here
+
+        // fetch default params from functionality
+        def defaultArgs = thisConfig.functionality.allArguments
+          .findAll { it.containsKey("default") }
+          .collectEntries { [ it.plainName, it.default ] }
+
+        // fetch overrides in params
+        def paramArgs = thisConfig.functionality.allArguments
+          .findAll { par ->
+            def argKey = key + "__" + par.plainName
+            params.containsKey(argKey) && params[argKey] != "viash_no_value"
+          }
+          .collectEntries { [ it.plainName, params[key + "__" + it.plainName] ] }
+        
+        // fetch overrides in data
+        def dataArgs = thisConfig.functionality.allArguments
+          .findAll { data_.containsKey(it.plainName) }
+          .collectEntries { [ it.plainName, data_[it.plainName] ] }
+        
+        // combine params
+        def combinedArgs = defaultArgs + paramArgs + processArgs.args + dataArgs
+
+        // remove arguments with explicit null values
+        combinedArgs.removeAll{it.value == null}
+
+        combinedArgs = processInputs(combinedArgs, thisConfig, id_, key)
+
+        [id_, combinedArgs] + tuple.drop(2)
+      }
+
+    out0_ = mid4_
+      | _debug(processArgs, "processed")
+      // run workflow
+      | innerWorkflowFactory(processArgs)
+      // check output tuple
+      | map { id_, output_ ->
+        output_ = processOutputs(output_, thisConfig, id_, key)
+
+        if (processArgs.auto.simplifyOutput && output_.size() == 1) {
+          output_ = output_.values()[0]
+        }
+
+        [id_, output_]
+      }
+
+    // TODO: this join will fail if the keys changed during the innerWorkflowFactory
+    // join the output [id, output] with the previous state [id, state, ...]
+    out1_ = out0_.join(mid2_, failOnDuplicate: true)
+      // input tuple format: [id, output, prev_state, ...]
+      // output tuple format: [id, new_state, ...]
+      | map{
+        def new_state = processArgs.toState(it)
+        [it[0], new_state] + it.drop(3)
+      }
+      | _debug(processArgs, "output")
```

Comment:
I just realised this join will fail if the `innerWorkflowFactory` changes the ids (cc @tverbeiren: cfr that one discussion we had).